jerry-core/parser/js/opcodes-dumper.h

```diff
@@ -16,25 +16,238 @@
 #ifndef OPCODES_DUMPER_H
 #define OPCODES_DUMPER_H
 
-#include "opcodes.h"
 #include "ecma-globals.h"
 #include "lexer.h"
+#include "opcodes.h"
+#include "serializer.h"
 
-typedef enum __attr_packed___
+/**
+ * Operand (descriptor of value or reference in context of parser)
+ */
+class jsp_operand_t
```

Comment:
@akiss77, I've performed the measurements on the following versions (if I didn't missed anything, the versions are only different usage of struct / class for parser `operand`):

https://github.com/Samsung/jerryscript/compare/Ruben-class-usage-investigation-prev...Ruben-class-usage-investigation-new?expand=1

Performance was measured in `--parse-only` mode, as the changes involve only parser.
Benchmark, used for the test was generated by the following script:
```
rm -f bench.js
for i in `seq 1 1 100`
do
    echo "a = 1 + 1 + 1 + 1 ? 1 * 1 * 1 : 1 & (1 * a * b + o.p * q - 1 + 'str');"
    echo "if (a) {b = 1 + 1 + 1 + 1 ? 1 * 1 * 1 : 1 ^ (1 * a * b + o.p * q - 1 + 'str');}"
    echo "else { for (i = 0; i < 1000; i++) { c = 1 + 1 + 1 + 1 ? 1 * 1 * 1 : 1 | (1 * a * b + o.p * q - 1 + 'str'); } }"
done >> bench.js
```

- release.mcu_stm32f4-cp (gcc version 4.8.2 (4.8.2-14ubuntu1+6)):
  - binary size
    - struct: 150700 bytes
    - class: 151012 bytes

- release.linux (gcc version 4.8.2 (Ubuntu/Linaro 4.8.2-16ubuntu4), hard float):
  - binary size (after `strip`):
    - struct: 190152 bytes
    - class: 190152 bytes
  - performance (Samsung Exynos 5250 Arndale Board, ARMv7, Cortex A15, 1GHz; every value in the sequence is a mean by 5 consecutive measurements - using script `tools/perf.sh`; second CPU core used for measurements - using `taskset -c 1`):
    - struct:
      - 3.332 (core temperature before - 34000 C, after - 39 C)
      - 3.328 (core temperature before - 39000 C, after - 39 C)
      - 3.334 (core temperature before - 40000 C, after - 40 C)
      - 3.34 (core temperature before - 40000 C, after - 40 C)
      - 3.334 (core temperature before - 40000 C, after - 40 C)
    - class:
      - 3.331 (core temperature before - 39 C, after - 39 C)
      - 3.334 (core temperature before - 39 C, after - 40 C)
      - 3.33 (core temperature before - 40 C, after - 40 C)
      - 3.337 (core temperature before - 40 C, after - 40 C)
      - 3.334 (core temperature before - 40 C, after - 40 C)

On mcu version binary size of build with class-based implementation is greater on 312 bytes.
Seems that this is caused by extra inline of a functions in build with C++ class-based implementation, relatively to build with C struct-based implementation.

In the performance measurements above share of `parse_*` functions was very small - about 0.01% or less.

To increase share of `parse_*` functions, I've added some performance optimizations (for heap allocator and linked list collection) to both builds.
On the new versions I've got the following (same environment; each value is mean of 20 measurements):
  - struct:
    - 0.1445 (core temperature before - 34 C, after - 37 C)
    - 0.14475 (core temperature before - 38 C, after - 38 C)
    - 0.14475 (core temperature before - 38 C, after - 39 C)
    - 0.14475 (core temperature before - 39 C, after - 39 C)
    - 0.14475 (core temperature before - 39 C, after - 39 C)
  - class:
    - 0.14525 (core temperature before - 37 C, after - 38 C)
    - 0.145 (core temperature before - 38 C, after - 38 C)
    - 0.14475 (core temperature before - 39 C, after - 39 C)
    - 0.145 (core temperature before - 39 C, after - 39 C)
    - 0.14525 (core temperature before - 39 C, after - 39 C)

So, time of build that uses class implementation is slower on about 0.2%.

However, the following is part of performance profile, consisting of first two `parse_*expression` functions in each profile, collected using `perf record` utility with frequency argument `-F 50000`:
- struct:
  - parse_member_expression: 0.36%
  - parse_unary_expression: 0.27%
- class:
  - parse_member_expression: 0.35% (a bit less)
  - parse_unary_expression: 0.27% (same)

Considering the profile, we can see that actually reduce of performance is not relevant to change of `parse_*expression` functions' performance.

For example, in the same profile:
- struct:
  - serializer_dump_op_meta: 0.22%
  - generate_instr: 0.29%
  - current_locus: 0.22%
  - linked_list_element: 12.86%
  - memcpy: 5.23%
- class:
  - serializer_dump_op_meta: 0.16%
  - generate_instr: 0.37%
  - current_locus: 0.49%
  - linked_list_element: 13.33%
  - memcpy: 5.22%

Considering the above, in the case the C++-feature itself doesn't involve binary size or performance.

For me it seems that, the C++-feature itself doesn't involve binary size of performance in general case too, and so can be used, with accurate consideration of consequences in each case (as for any other way of implementation).

In other words, there definitely can be some cases, in which performance is reduced because of using some feature, but it seems to me that the cases should be considered in the feature-independent way, i.e. by analysing actual origin of the reduce.

Concerning classes, I don't see any "physical" difference between basic classes and structures. For me, it seems to be just different way to declare the same behaviour.
Actual difference can occur, for example, during implementation of some default operations (like default constructor) that would add unnecessary operations in some cases, in which the operations are not optimized out. The cases should definitely be considered very accurately.

@akiss77, @zherczeg, @egavrin, @sand1k, could you, please, share your opinion about this?